tests/rptest/tests/group_membership_test.py

```diff
@@ -272,62 +271,92 @@ def test_leadership_transfer(self):
+            if not metrics:
+                self.logger.debug("No metrics found")
                 return False
+            metrics = metrics.label_filter(dict(group=group)).samples
+            for metric in metrics:
+                self.logger.debug(
+                    f"Retrieved metric from node={metric.node.account.hostname}: {metric}"
+                )
             return all([
                 metric.node.account.hostname == node.account.hostname
-                for metric in metrics_offsets
+                for metric in metrics
             ])
 
-        def transfer_completed(new_leader_node):
-            return self.redpanda.nodes[get_group_leader() - 1].account.hostname \
-                    == new_leader_node.account.hostname
-
-        leader_node = self.redpanda.nodes[get_group_leader() - 1]
-        check_metric_from_node(leader_node)
-
-        # Check transfer leadership to another node
-        for i in range(3):
-            new_leader_node = random.choice(
-                list(filter(lambda x: x != leader_node, self.redpanda.nodes)))
-
-            admin.transfer_leadership_to(
-                namespace="kafka_internal",
-                topic="group",
-                partition=0,
-                target=self.redpanda.idx(new_leader_node))
-
-            wait_until(lambda: transfer_completed(new_leader_node) and
-                       check_metric_from_node(new_leader_node),
+        def transfer_leadership(new_leader):
+            """
+            Request leadership transfer of the internal consumer group partition
+            and check that it completes successfully.
+            """
+            self.logger.debug(
+                f"Transferring leadership to {new_leader.account.hostname}")
+            admin.transfer_leadership_to(namespace="kafka_internal",
+                                         topic="group",
+                                         partition=0,
+                                         target=self.redpanda.idx(new_leader))
+            for _ in range(3):  # re-check a few times
+                leader = self.redpanda.get_node(get_group_leader())
+                if leader == new_leader:
+                    return True
+                time.sleep(1)
+            return False
+
+        def partition_ready():
+            """
+            All replicas present and known leader
+            """
+            partition = get_group_partition()
+            self.logger.debug(f"XXXXX: {partition}")
+            return len(
+                partition['replicas']) == 3 and partition['leader_id'] >= 0
+
+        def select_next_leader():
+            """
+            Select a leader different than the current leader
+            """
+            wait_until(partition_ready, timeout_sec=30, backoff_sec=5)
+            partition = get_group_partition()
+            replicas = partition['replicas']
+            assert len(replicas) == 3
```

Comment:
[nit]
Isn't this check excessive?
We have already checked it in `partition_ready`

Reply:
the data comes from two separate calls into redpanda, so in principle it could hit a different broker that had out of date information.

you are right that it is probably excessive, but it's tests so lots of checks on assumptions can be helpful :)